code/modules/station_goals/bluespace_tap.dm

```diff
@@ -255,6 +255,75 @@
 	if(!powernet)
 		connect_to_network()
 
+/obj/machinery/power/bluespace_tap/update_icon_state()
+	. = ..()
+
+	if(length(active_nether_portals))
+		icon_state = "redspace_tap"
+		return
+
+	if(get_available_power() <= 0)
+		icon_state = base_icon_state
+	else
+		icon_state = "[base_icon_state][get_icon_state_number()]"
+
+
+/obj/machinery/power/bluespace_tap/update_overlays()
+	. = ..()
+
+	overlays.Cut()
+	underlays.Cut()
+
+	if(length(active_nether_portals))
+
+		overlays += icon(icon, "redspace")
+		set_light(15, 5, "#ff0000")
+		return
+
+	if(stat & (BROKEN|NOPOWER))
+		set_light(0)
+	else
+		set_light(1, 1, "#353535")
+
+	if(get_available_power())
+		overlays += icon(icon, "screen")
+		if(light)
+			underlays += emissive_appearance(icon, "light_mask")
+
+
+/obj/machinery/power/bluespace_tap/proc/get_icon_state_number()
+	switch(input_level)
+		if(0)
+			return 0
+		if(1 to 2)
+			return 1
+		if(3 to 5)
+			return 2
+		if(6 to 7)
+			return 3
+		if(8 to 10)
+			return 4
+		if(11 to INFINITY)
+			return 5
+
+/obj/machinery/power/bluespace_tap/power_change()
+	. = ..()
+	if(stat & (BROKEN|NOPOWER))
+		set_light(0)
+	else
+		set_light(1, 1, "#353535")
+	if(.)
+		update_icon(UPDATE_ICON)
```

Comment:
`update_icon()` by default uses `UPDATE_ICON`, so you don't need to put that everywhere. What you want to use is either `UPDATE_ICON_STATE` or `UPDATE_OVERLAYS` if you know of any situation where only one of the two is being affected.

This prevents unnecessary updating of the icon state or overlays unless you're updating both on purpose.

Reply:
@S34NW